### PR TITLE
Allow binaries as parts of logger_formatter template

### PR DESCRIPTION
**Pull request by hauleth**
This small change will simplify integration of Erlang's logger with
Elixir applications where users are more accustomed to usage of binaries
instead of char lists as a "string" type.

### CODE REVIEW

File: lib/kernel/src/logger_formatter.erl [hunk 1]
```diff
@@ -531,6 +531,9 @@ check_template([Str|T]) when is_list(Str) ->
         true -> check_template(T);
         false -> error
     end;
+check_template([Bin|T]) when is_binary(Bin) ->
+    Str = binary:bin_to_list(Bin),
```
Comment (ferd) [inline lib/kernel/src/logger_formatter.erl]: this wouldn't be unicode-safe compared to `unicode:characters_to_list/2-3`.

Comment (hauleth) [reply lib/kernel/src/logger_formatter.erl]: Fixed.

**Comment by garazdawi** [comment]
The `string()` type should probably just be changed to `chardata()` instead of adding `binary()`. However, that means that we should add tests for deep `chardata()` in all the places. i.e.

```
Template = [{metakey,[[$e,<<"x">>,$i,[<<"s">>],":"],metakey],[[[[[[[[["does ",<<"not">>," exist"]]]]]]]]]},[[[<<"more">>]],"deep"]].
```

**Comment by hauleth** [comment]
@garazdawi added such test as well.

File: lib/kernel/src/logger_formatter.erl [hunk 2]
```diff
@@ -35,7 +35,7 @@
                     template        => template(),
                     time_designator => byte(),
                     time_offset     => integer() | [byte()]}.
--type template() :: [metakey() | {metakey(),template(),template()} | string()].
```
Comment (garazdawi) [inline lib/kernel/src/logger_formatter.erl]: The type should be:

    -type template() :: [metakey() | {metakey(),template(),template()} | unicode:chardata()].

**Comment by garazdawi** [comment]
Better late than never! :) Thanks!

**Pull request merged**